wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,350 @@
+from __future__ import annotations
+from typing import Iterator, List, Tuple, Optional, BinaryIO
+
+from .at import at
+from .casts import CastableType, to_clvm_object, int_from_bytes, int_to_bytes
+from .chia_dialect import CHIA_DIALECT
+from .clvm_rs import run_serialized_chia_program
+from .clvm_storage import CLVMStorage
+from .clvm_tree import CLVMTree
+from .curry_and_treehash import CurryTreehasher
+from .eval_error import EvalError
+from .replace import replace
+from .ser import sexp_from_stream, sexp_to_stream, sexp_to_bytes
+from .tree_hash import sha256_treehash
+
+
+
+class Program(CLVMStorage):
+    """
+    A wrapper around `CLVMStorage` providing many convenience functions.
+    """
+
+    UNSAFE_MAX_COST: Optional[int] = None
+
+    curry_treehasher: CurryTreehasher = CurryTreehasher(CHIA_DIALECT)
+    _cached_serialization: Optional[bytes]
+
+    # serialization/deserialization
+
+    @classmethod
+    def parse(cls, f: BinaryIO) -> Program:
+        return sexp_from_stream(f, cls.new_pair, cls.new_atom)
+
+    def stream(self, f: BinaryIO) -> None:
+        sexp_to_stream(self, f)
+
+    @classmethod
+    def from_bytes(cls, blob: bytes, calculate_tree_hash: bool = True) -> Program:
+        obj, cursor = cls.from_bytes_with_cursor(
+            blob, 0, calculate_tree_hash=calculate_tree_hash
+        )
+        return obj
+
+    @classmethod
+    def from_bytes_with_cursor(
+        cls, blob: bytes, cursor: int, calculate_tree_hash: bool = True
+    ) -> Tuple[Program, int]:
+        tree = CLVMTree.from_bytes(
+            blob[cursor:], calculate_tree_hash=calculate_tree_hash
+        )
+        obj = cls.wrap(tree)
+        new_cursor = len(bytes(tree)) + cursor
+        return obj, new_cursor
+
+    @classmethod
+    def fromhex(cls, hexstr: str) -> Program:
+        return cls.from_bytes(bytes.fromhex(hexstr))
+
+    def __bytes__(self) -> bytes:
+        if self._cached_serialization is None:
+            self._cached_serialization = sexp_to_bytes(self)
+        if not isinstance(self._cached_serialization, bytes):
```

Comment:
I'm a bit confused by this question, as that line of code appears verbatim in the above.

Reply:
in what case is `self._cached_serialization` anything but `bytes`?
the type of `_cached_serialization` is `Optional[bytes]`. The only two types it can have is `None` or `bytes`. So, only if `sexp_to_bytes()` return `None` would you enter this if-block. But a more intuitive way of saying that would be `if self._cached_serialization is not None`

Reply:
The original idea is that in some cases it may be `memoryview` because that has the nice property that slices are cheap (whereas `bytes` slices do a copy). It's been a while since I looked at this so I'm not sure offhand if there are benefits to this approach and in what cases they occur.

Reply:
I see. so `_cached_serialization` should probably have the type: `Optional[Union[bytes, memoryview]]` then. and I would expect `sexp_to_bytes()` would return `Union[bytes, memoryview]`, or something like that